contracts/cw-named-groups/src/tests.rs

```diff
@@ -0,0 +1,630 @@
+use std::{collections::HashSet, iter::FromIterator};
+
+use crate::{
+    msg::{DumpResponse, ExecuteMsg, Group, InstantiateMsg, QueryMsg},
+    ContractError,
+};
+use cosmwasm_std::{Addr, Empty, StdError, StdResult};
+use cw_multi_test::{App, Contract, ContractWrapper, Executor};
+
+fn named_group_contract() -> Box<dyn Contract<Empty>> {
+    let contract = ContractWrapper::new(
+        crate::contract::execute,
+        crate::contract::instantiate,
+        crate::contract::query,
+    );
+    Box::new(contract)
+}
+
+const ADMIN: &str = "DAO";
+const USER1: &str = "USER1";
+const USER2: &str = "USER2";
+
+fn group_factory(id: u8) -> Group {
+    Group {
+        name: format!("GROUP{}_NAME", id),
+        addresses: HashSet::from_iter(vec![format!("USER{}", id)]),
+    }
+}
+
+fn mock_app() -> App {
+    App::default()
+}
+
+fn instantiate(groups: Option<Vec<Group>>) -> Result<(App, Addr), ContractError> {
+    let mut app = mock_app();
+    let contract_id = app.store_code(named_group_contract());
+
+    let msg = InstantiateMsg { groups };
+    let instantiation_result =
+        app.instantiate_contract(contract_id, Addr::unchecked(ADMIN), &msg, &[], "test", None);
+
+    if let Ok(contract_addr) = instantiation_result {
+        Ok((app, contract_addr))
+    } else {
+        Err(instantiation_result.unwrap_err().downcast().unwrap())
+    }
+}
+
+fn dump(app: &App, contract_addr: &Addr) -> DumpResponse {
+    app.wrap()
+        .query_wasm_smart(contract_addr, &QueryMsg::Dump {})
+        .unwrap()
+}
+
+mod instantiate {
+    use super::*;
+
+    #[test]
+    fn instantiate_with_no_groups() {
+        let (app, contract_addr) = instantiate(None).unwrap();
+
+        // Ensure there are no groups.
+        let dump_result = dump(&app, &contract_addr);
+        assert_eq!(dump_result.groups.len(), 0);
+    }
+
+    #[test]
+    fn instantiate_with_groups() {
+        let groups = vec![group_factory(1), group_factory(2), group_factory(3)];
+
+        let (app, contract_addr) = instantiate(Some(groups.clone())).unwrap();
+
+        // Ensure there are the expected groups.
+        let dump_result = dump(&app, &contract_addr);
+        assert_eq!(dump_result.groups, groups);
```

Comment:
Let's test that the `ADDRESSES` map is properly instantiated here as well. Don't think that works at the moment.

Reply:
What do you mean? `GROUPS.save` on line 58 is only called if there are any initial groups, and `ADDRESSES.save` is the same. It doesn't seem like there's any function to instantiate an empty Map, as `.save` and `.update` take keys and values to store. The API makes it opaque if the underlying data store has been instantiated or not, but I assume just the `Map::new` in `state.rs` takes care of that. What am I missing?

Reply:
I understand what you mean! Another test should be stronger, not this one. Resolving this because I added more code to verify both `List...` queries return the right data after instantiating with initial groups.